hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java

```diff
@@ -252,13 +252,10 @@ public DBCheckpoint getCheckpoint(Path tmpdir, boolean flush)
       checkpoint = getDbStore().getCheckpoint(flush);
     } finally {
       // Unpause the compaction threads.
-      synchronized (getDbStore().getRocksDBCheckpointDiffer()) {
-        differ.decrementTarballRequestCount();
-        differ.notifyAll();
-        long elapsedTime = System.currentTimeMillis() - startTime;
-        LOG.info("Compaction pausing {} ended. Elapsed ms: {}",
-            pauseCounter, elapsedTime);
-      }
+      differ.decrementTarballRequestCount();
+      differ.notifyAll();
```

Comment:
`notifyAll()` needs the object lock (see [docs](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#notifyAll--).  BTW it should be `synchronized (differ)`, which is both simpler and safer.

`differ.decrementTarballRequestCount()` may or may not need the lock, I cannot comment on that.

`LOG.info` should be outside the lock.

Reply:
I remember that `notifyAll()` needs the object lock but I forgot the reason https://github.com/apache/ozone/pull/5070/files#r1264234029. Thanks for  pointing to the doc.

IDE raises "Synchronization on local variable 'differ'" warning if it is done like `synchronized (differ)`. That's the reason it was changed to `synchronized (getDbStore().getRocksDBCheckpointDiffer())`.

BTW I don't think it matter if lock is taken like `synchronized (differ)` or `synchronized (getDbStore().getRocksDBCheckpointDiffer())`. It is still at instance level and can cause the deadlock if I'm not wrong.

`differ.decrementTarballRequestCount()` and `LOG.info` don't need the lock. decrementTarballRequestCount updates AtomicInteger so lock is not needed.

Reply:
To fix the deadlock, I moved `notifyAll()` inside a synchronized block and removed the nested locks in `RocksDBCheckpointDiffer#pruneSstFiles`. RocksDBCheckpointDiffer's instance lock is only needed to know which sstFiles can be removed. Once that's know, actual removal has to be inside `BootstrapStateHandler#lock`.

Reply:
> IDE raises "Synchronization on local variable 'differ'" warning if it is done like `synchronized (differ)`. That's the reason it was changed to `synchronized (getDbStore().getRocksDBCheckpointDiffer())`.
> 
> BTW I don't think it matter if lock is taken like `synchronized (differ)` or `synchronized (getDbStore().getRocksDBCheckpointDiffer())`. It is still at instance level and can cause the deadlock if I'm not wrong.

The reason for using `synchronized (differ)` is to ensure we have the lock on the same instance on which we call `notifyAll`.  I didn't mean it would solve the deadlock.

Better solution for the IDE warning: move `notifyAll()` into `RocksDBCheckpointDiffer#decrementTarballRequestCount`, and wrap it in a `synchronized (this)` block there.

Reply:
It makes sense to more it to `RocksDBCheckpointDiffer#decrementTarballRequestCount`. Updated as per the suggestion. Thanks @adoroszlai.